src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -705,20 +683,28 @@ def setUp(self):
         self.manager_class = self.params.get("manager_class", "/", "Orterun")
 
         # Determine which hosts to use as servers and optionally clients.
-        self.hostlist_servers = self.get_hosts_from_yaml(
-            "test_servers", "server_partition", "server_reservation", "/run/hosts/*")
-        self.hostlist_clients = self.get_hosts_from_yaml(
-            "test_clients", "client_partition", "client_reservation", "/run/hosts/*")
-
-        # Optionally remove any servers that may have ended up in the client list.  This can occur
-        # with tests using slurm partitions as they are setup with all hosts.
-        if self.slurm_exclude_servers:
-            self.log.debug(
-                "Excluding any %s servers from the current client list: %s",
-                self.hostlist_servers, self.hostlist_clients)
-            new_client_list = self.hostlist_clients.difference(self.hostlist_servers)
-            self.slurm_exclude_nodes = self.hostlist_clients.difference(new_client_list)
-            self.hostlist_clients = new_client_list
+        info = HostInfo()
+        info.set_test_hosts(self)
+        self.hostlist_servers = info.servers
+        self.hostlist_clients = info.clients
+        self.server_partition = info.server_partition.name
+        self.client_partition = info.client_partition.name
+        self.server_reservation = info.server_partition.reservation
+        self.client_reservation = info.client_partition.reservation
```

Comment:
This pattern seems a little strange. We define an independent object `HostInfo`. Then update that object using `self`. Then update `self` using the object. So there is a sort of recursive dependency on these objects.

I'm not sure how to do this better without refactoring all code to use `self.info.*` instead of setting all those `self.* = info.*`